core/observers/STFLogger.cpp

```diff
@@ -13,50 +13,95 @@ namespace pegasus
         stf_writer_.addTraceInfo(stf::TraceInfoRecord(stf::STF_GEN::STF_TRANSACTION_EXAMPLE, 0, 0,
                                                       0, "Trace from Pegasus"));
 
-        if (reg_width == 64)
+        if (state->getXlen() == 32)
         {
-            stf_writer_.setHeaderIEM(stf::INST_IEM::STF_INST_IEM_RV64);
-            stf_writer_.setTraceFeature(stf::TRACE_FEATURES::STF_CONTAIN_RV64);
+            stf_writer_.setHeaderIEM(stf::INST_IEM::STF_INST_IEM_RV32);
         }
         else
         {
-            stf_writer_.setHeaderIEM(stf::INST_IEM::STF_INST_IEM_RV32);
+            stf_writer_.setHeaderIEM(stf::INST_IEM::STF_INST_IEM_RV64);
+            stf_writer_.setTraceFeature(stf::TRACE_FEATURES::STF_CONTAIN_RV64);
         }
 
+        // Is there a parameter for the user to enable page tables and physical memory tracing?
+        stf_writer_.setTraceFeature(stf::TRACE_FEATURES::STF_CONTAIN_PTE);
+
         stf_writer_.setISA(stf::ISA::RISCV);
         stf_writer_.setHeaderPC(inital_pc);
         stf_writer_.finalizeHeader();
-        (void)state; // Unused parameter, but will be implemented when recording inital state of
-                     // registers
-        // recordRegState_(state); record inital state of registers
+        recordRegState_(state);
     }
 
-    void STFLogger::writeInstruction_(const PegasusInst* inst)
+    // METHODS
+    void STFLogger::postExecute_(PegasusState* state)
     {
-        if (fault_cause_.isValid() || interrupt_cause_.isValid())
+        if (state->getCurrentInst()->getOpcodeSize() == 2)
         {
-            return;
+            stf_writer_ << stf::InstOpcode16Record(state->getCurrentInst()->getOpcode());
+        }
+        else
+        {
+            stf_writer_ << stf::InstOpcode32Record(state->getCurrentInst()->getOpcode());
         }
 
```

Comment:
The InstOpcode records mark the end of an instruction group and should be written after any other records you want attached to the instruction.